align/pnr/grid_constraints.py

```diff
@@ -0,0 +1,128 @@
+import math
+from itertools import product
+from functools import reduce
+from collections import defaultdict
+from .hpwl import compute_topoorder
+import logging
+
+from ..schema.constraint import OffsetsScalings, PlaceOnGrid
+
+logger = logging.getLogger(__name__)
+
+def check(r, x, scaling):
+    return any((x % r.pitch) in ored_term.offsets and
+               scaling in ored_term.scalings
+               for ored_term in r.ored_terms)
+
+def simplify(pairs):
+    offsets_with_pos_scaling = {o for o, s in pairs if s == 1}
+    offsets_with_neg_scaling = {o for o, s in pairs if s == -1}
+    offsets_with_both_scalings = offsets_with_pos_scaling.intersection(offsets_with_neg_scaling)
+    offsets_with_pos_scaling -= offsets_with_both_scalings
+    offsets_with_neg_scaling -= offsets_with_both_scalings
+    if offsets_with_both_scalings:
+        yield OffsetsScalings(offsets=sorted(offsets_with_both_scalings), scalings=[1, -1])
+    if offsets_with_pos_scaling:
+        yield OffsetsScalings(offsets=sorted(offsets_with_pos_scaling), scalings=[1])
+    if offsets_with_neg_scaling:
+        yield OffsetsScalings(offsets=sorted(offsets_with_neg_scaling), scalings=[-1])
+
+
+def lcm(*xs):
+    def lcm2(a, b):
+        return (a*b) // math.gcd(a, b)
+    assert xs
+    if len(xs) == 1:
+        return xs[0]
+    else:
+        return reduce(lcm2, xs)
+
+
+def merge(*rs):
+    new_pitch = lcm(*(r.pitch for r in rs))
+    new_direction = rs[0].direction
+    assert all(new_direction == r.direction for r in rs)
+
+    def gen_pairs(r):
+        for coarse_offset in range(0, new_pitch, r.pitch):
+            for ored_term in r.ored_terms:
+                for offset, scaling in product(ored_term.offsets, ored_term.scalings):
+                    yield (coarse_offset + offset, scaling)
+
+    pairs = set.intersection(*(set(gen_pairs(r)) for r in rs))
+    res = PlaceOnGrid(direction=new_direction, pitch=new_pitch, ored_terms=list(simplify(pairs)))
+
+    if not res.ored_terms:
+        logger.error(f'No legal grid positions: {res}')
+
+    return res
+
+def split_directions_and_merge(*rs):
+    split_directions = defaultdict(list)
+    for r in rs:
+        split_directions[r.direction].append(r)
+
+    return [merge(*v) for v in split_directions.values()]
+
+
+def gen_constraints_for_module(m, modules, leaves):
+    pog_constraints = []
+    for instance in m['instances']:
+        ctn = instance['concrete_template_name']
+        constraints = None
+        # I don't know why I have to do this: I can't check to see if 'constraints' is in leaves[ctn] (it never is.)
+        if ctn in leaves:
+            try:
```

Comment:
@parijatm Can you think of any reason I have to use this try except block here? When I try to do an explicit membership test (`'constraints' in leaves[ctn]`) it always returns `False`. I can't use the `get` method either

Reply:
I think the answer is in `hacks.py`. `VerilogJson*` classes are subclasses of `DictEmulator` which is not a subclass of dict. That is why it is tricky to work with these objects which give the impression that they are dictionaries.